### PR TITLE
WIP: Storage-ng based AutoYaST partitioning cloning

### PR DESCRIPTION
**Pull request by ancorgs**
Part of https://trello.com/c/mAFPlMj7/602-5-autoyast-ng-partitioner-clone-the-system-basic

Used from https://github.com/yast/yast-autoinstallation/pull/311

Still missing:

 * Unit tests
 * Changelog and version

### CODE REVIEW

File: src/lib/y2storage/autoinst_profile/drive_section.rb [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+      def initialize
+        @partitions = []
+        @skip_list = SkipListSection.new([])
+      end
+
+      # Method used by {.new_from_hashes} to populate the attributes.
+      #
+      # It only enforces default values for #type (:CT_DISK) and #use ("all")
+      # since the {AutoinstProposal} algorithm relies on them.
+      def init_from_hashes(hash)
```
Comment (imobachgs) [inline src/lib/y2storage/autoinst_profile/drive_section.rb]: np: document `hash` argument.

File: src/lib/y2storage/autoinst_profile/drive_section.rb [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+        super
+        @type ||= :CT_DISK
+        @use  ||= "all"
+        @partitions = partitions_from_hash(hash)
+        @skip_list = SkipListSection.new_from_hashes(hash.fetch("skip_list", []))
+      end
+
+      # Clones a drive into an AutoYaST profile section, creating an instance
+      # if this class from the information in a block device.
```
Comment (imobachgs) [inline src/lib/y2storage/autoinst_profile/drive_section.rb]: Sorry, I do not understand the second line.

File: src/lib/y2storage/autoinst_profile/drive_section.rb [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+        return false if disk.partitions.empty?
+
+        @type = :CT_DISK
+        @device = Yast::Arch.s390 ? disk.udev_full_paths.first : disk.name
+        @disklabel = disk.partition_table.type.to_s
+
+        @partitions = partitions_from_disk(disk)
+        return false if @partitions.empty?
+
+        if reuse_partitions?(disk)
```
Comment (imobachgs) [inline src/lib/y2storage/autoinst_profile/drive_section.rb]: This behaviour is really tricky. Thanks for documenting it :)

File: src/lib/y2storage/autoinst_profile/drive_section.rb [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+      end
+
+      # Whether AutoYaST considers a partition to be part of a Windows
+      # installation and not directly relevant for the system being
+      # cloned.
+      #
+      # NOTE: to ensure backward compatibility, this method implements the
+      # logic present in the old AutoYaST exporter (that used to live in
+      # AutoinstPartPlan#ReadHelper). Check the comments in the code to know
+      # more about the what is checked and why.
```
Comment (imobachgs) [inline src/lib/y2storage/autoinst_profile/drive_section.rb]: Please, add a link, we expect that code to be removed soon.

File: src/lib/y2storage/autoinst_profile/drive_section.rb [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+        # commit 795a18a795cd45d7e5f4d (January 2017) in order to fix
+        # bsc#192342. The PPC bootloader was switching the id of the partition
+        # from id 0x41 (PReP) to id 0x06 (FAT16) and as a result the AutoYaST
+        # exporter was ignoring the partition (considering it to be a Windows
+        # partition). Very likely, the intention of the fix was just to stop
+        # considering such FAT16+boot partitions as part of Windows.
+        # Unfortunately, the introduced fix affected all Windows-related ids,
+        # not only FAT16.
+        # That side effect has been there for 10+ years, so let's keep it.
+        return false if partition.boot?
```
Comment (imobachgs) [inline src/lib/y2storage/autoinst_profile/drive_section.rb]: What about just `!partition.boot?` ?

File: src/lib/y2storage/autoinst_profile/drive_section.rb [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+      #   a Windows system (see #windows?)
+      def skip_partition?(partition)
+        partition.type.is?(:extended) || windows?(partition)
+      end
+
+      # Whether all partitions in the drive should have "create" set to false
+      # (so no new partitions will be actually created in the target system).
+      #
+      # NOTE: This implements logic that was present in the old exporter and
+      # returns true if there is a Windows partition (see {.windows?}) that is
```
Comment (imobachgs) [inline src/lib/y2storage/autoinst_profile/drive_section.rb]: `#windows?`

File: src/lib/y2storage/autoinst_profile/drive_section.rb [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+      # placed in the disk after any non-Windows partition.
+      #
+      # @param disk [Y2Storage::Partitionable]
+      # @return [Boolean]
+      def reuse_partitions?(disk)
+        linux_already_found = false
+        disk.partitions.each do |part|
+          if windows?(part)
+            return true if linux_already_found
```
Comment (imobachgs) [inline src/lib/y2storage/autoinst_profile/drive_section.rb]: scary... :)

File: src/lib/y2storage/autoinst_profile/partition_section.rb [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+
+      # @!attribute loop_fs
+      #   @return [Boolean] undocumented attribute
+
+      def initialize
+        @subvolumes = []
+      end
+
+      # Clones a device into an AutoYaST profile section, creating an instance
+      # if this class from the information in a partition or LVM logical volume.
```
Comment (imobachgs) [inline src/lib/y2storage/autoinst_profile/partition_section.rb]: Sorry, I do not get the second line.

File: src/lib/y2storage/autoinst_profile/partition_section.rb [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+      end
+
+      # Whether the given existing partition should be reported as GRUB (GPT
+      # Bios Boot) in the cloned profile.
+      #
+      # @note To ensure backward compatibility, this method implements the
+      # logic present in the old AutoYaST exporter (that used to live in
+      # AutoinstPartPlan#ReadHelper). So it returns true for any partition with
+      # a Windows-related ID that is configured to be mounted in /boot*
+      # See commit 54e236cd428636b3bf8f92d2ac2914e5b1d67a90
```
Comment (imobachgs) [inline src/lib/y2storage/autoinst_profile/partition_section.rb]: A link would be nice.

File: src/lib/y2storage/autoinst_profile/section_with_attributes.rb [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+
+      def attributes
+        self.class.attributes
+      end
+
+      # Whether an attribute must be skipped during import/export.
+      #
+      # @return [Boolean] true is the value is blank
+      def attrib_skip?(value)
+        return true if value.nil?
```
Comment (imobachgs) [inline src/lib/y2storage/autoinst_profile/section_with_attributes.rb]: np: why three different lines?

File: src/lib/y2storage/proposal/autoinst_devices_planner.rb [hunk 11]
```diff
@@ -50,9 +50,9 @@ def initialize(devicegraph)
       def planned_devices(drives_map)
         result = []
 
-        drives_map.each_pair do |disk_name, drive_spec|
```
Comment (imobachgs) [inline src/lib/y2storage/proposal/autoinst_devices_planner.rb]: It will conflict with #276 😱

**Comment by ancorgs** [comment]
Closed in favor of #284 (which includes rebasing and fix for the mentioned issues).

**Pull request closed without merging**